### DISCUSSION TITLE
Failing tests on async exceptions from other threads/dispatchers?

### DISCUSSION THREAD

**Discussion by egil** [xunit/xunit#2270] [Question] [answered]
I have a scenario with [bUnit](https://bunit.egilhansen.com/), where a user will create test for a component, that has asynchronous code which throws an exception. However, that exception should ideally be surfaced in the test, but it is not. Let me try to explain what is going on:

1. xUnit fact/test code news up a bUnit's TestRenderer and asks it to render a component, referred to as the component under test (CUT)
2. CUT performs an async operation, e.g. a HttpClient call.
3. When the async operation completes, the continuation is invoked on the TestRenderer's dispatcher, and any uncaught exceptions is captured by it and wont show up in the test, e.g. an exception from an HttpClient call.

When an unhandled exception is caught by the TestRenderer's dispatcher, a callback is invoked from where I could forward the exception to the xUnit fact/test. But how?
When I new-up the TestRenderer, I can access xUnits SynchronizationContext. But can I use that to break the current test?

Thanks, Egil

**Comment by bradwilson** [comment, marked as answer]
The only way you can realistically get an exception back to xUnit.net is to ensure that it is thrown on the thread that the system believes is running the test. This means you'll need a way to marshal and re-throw the exception back on the original thread.

I describe the catch/re-throw process here: https://twitter.com/bradwilson/status/1353588311940427776

The exercise left for the reader (aka, you 😄) is determining how to get that exception back onto the right thread to re-throw it.

**Reply by egil** [reply]
Thanks for the input Brad. Im already using the ExceptionDispatcher trick, think I saw your tweet before :)

My take is, that the only way to really push an exception back to the test thread, is to check during bUnit's TestContext's Dispose method, whether or not an uncaught exception hasn't been seen by the test, and then throw at that point. I know its not best practice throw during Dispose, but I think it makes sense in this case.
